### ISSUE TITLE
Ignore node_modules not working

### ISSUE THREAD

**Issue by dottodot** [TypeStrong/grunt-ts#318]
Sorry please bear me as not had any experience with typescript before but trying to compile a simple angular2 app but no matter what I try `grunt-ts` always compiles the node_modules as well and then throws a load of errors like

```
Duplicate identifier 'onloadstart'.
/Users/marcus/hosts/sails-angular/assets/node_modules/typescript/lib/lib.webworker.d.ts(1027,5): error TS2300: Duplicate identifier 'onprogress'.
/Users/marcus/hosts/sails-angular/assets/node_modules/typescript/lib/lib.webworker.d.ts(1028,5): error TS2300: Duplicate identifier 'ontimeout'.
/Users/marcus/hosts/sails-angular/assets/node_modules/typescript/lib/lib.webworker.d.ts(1115,5): error TS2300: Duplicate identifier 'type'.
/Users/marcus/hosts/sails-angular/assets/node_modules/typescript/lib/lib.webworker.d.ts(1116,5): error TS2300: Duplicate identifier 'end
```

Here's my config

```
module.exports = function(grunt) {
  grunt.config.set('ts', {
    compile: {
      files: [{
        src: [
          'assets/app/**/*.ts',
          "!assets/node_modules/**"
        ],
        dest: ''
      }],
      tsconfig: true,
      verbose: true
    }
  });

  grunt.loadNpmTasks('grunt-ts');
};
```

and I've also tried other not rules such as

```
!assets/node_modules/**/*.ts 
```

**Comment by basarat** [comment]
>  no matter what I try grunt-ts always compiles the node_modules as well and then throws a load of errors like

This is because of TypeScript's new node_modules resolution. If you don't want this to happen use the `moduleResoution` set to `classic` in your tsconfig.json :rose: http://json.schemastore.org/tsconfig

**Comment by basarat** [comment]
Feel free to ask further if its not clear :rose:

**Comment by dottodot** [comment]
I'm still getting the same result with it set to classic.

```
{
  "compilerOptions": {
    "target": "ES5",
    "module": "system",
    "moduleResolution": "classic",
    "sourceMap": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "removeComments": false,
    "noImplicitAny": false
  },
  "exclude": [
    "node_modules"
  ]
}
```
and it's a good 30 sec before it shows the error I had before.

**Comment by basarat** [comment]
@dottodot sorry. Can you share a working project I can clone and investigate :rose:

**Comment by dottodot** [comment]
Yes here's the link

https://bitbucket.org/dottodotdesign/sails-angular

It's a SailsJs app so not sure how familiar you are with it. Anyway the grunt config file is in tasks/config. Let me know if need any more details.

**Comment by basarat** [comment]
@dottodot I did the following: 

* clone
* npm install
* grunt

However the errors I got are different. 

![image](https://cloud.githubusercontent.com/assets/874898/12248570/6926eee6-b90e-11e5-98a2-f31a0f92d54d.png)

Did I do something wrong?:rose:

**Comment by dottodot** [comment]
@basarat Sorry you need to run npm install on the assets directory as well.

**Comment by basarat** [comment]
I'm struggling to get it to work too. But then the whole build seems not simple and can't devote more time to it...sorry. Leaving it open for a hero :rose:

**Comment by basarat** [comment]
Most likely not a bug, but a configuration error that I cannot track down

**Comment by dottodot** [comment]
I think I might have resolved it anyway. Putting the tsconfig in the assets folder seems to work with moduleResoution set back to node as my grunt config as

```
module.exports = function(grunt) {

	grunt.config.set('ts', {
		compile: {
      files: [{
        src: [
          'assets/app/**/*.ts'
        ],
        dest: ''
      }],
      tsconfig: './assets/tsconfig.json',
      verbose: true
		}
	});

	grunt.loadNpmTasks('grunt-ts');
};
```

**Comment by Rouche** [comment]
I am struggling with the same problem. We trying to switch to a tsconfig.json so our Maven build (wich use grunt) will have same configuration as our IDE plugin.
I cannot use glob in tsconfig.json because the plugin dont support it.

I cannot find what im missing:

Our original config wich works perfectly:
```
        ts: {
            main: {
                src: [pathWebapp + '**/*.ts', '!' + pathWebapp + '**/*.d.ts'],

                options: {
                    sourceMap: true,
                    declaration: false,
                    removeComments: false,
                    module: 'amd'
                }
            }
        },
```

Now with tsconfig, wich add all files in node_modules
```
        ts: {
            main: {
                tsconfig: true,
                ignoreFiles: true,
                src: [pathWebapp + '**/*.ts', '!' + pathWebapp + '**/*.d.ts']
            }
        },
```
tsconfig.json:
```
{
	"compileOnSave": true,
	
    "compilerOptions": {
        "target": "ES5",
        "declaration": false,
        "removeComments": false,
        "sourceMap": true,
        "module": "amd"
    }
}
```

**Comment by nycdotnet** [comment]
Thanks for the reports, everyone.

@basarat I'm doing significant work on tsconfig support atm (the new functionality in TypeScript 2.x such as `include`), so this is likely some wonkiness.

To the rest of the folks in this thread - I will be looking into this today.  The best and most reliable way to get grunt-ts to behave identically to "what TypeScript would do" is to use the "passThrough" argument.  This basically disables anything clever that grunt-ts might try to do and it just calls TypeScript with the referenced tsconfig.json file.

```js
grunt.initConfig({
  ts: {
    mytarget: {
      tsconfig: {
        tsconfig: './SomeFolder/sometsconfig.json',
        passThrough: true
      }
    }
  }
});
```

Beyond that, grunt-ts tries to do the best it can with what you give it, and for example will union together any files found from the grunt `src` or `files` with what TypeScript would resolve based on the tsconfig.json `include`, `exclude`, and `files` properties.  Much of this code is brand new in the grunt-ts v6 betas, though, so there still could be some bugs.

**Comment by Rouche** [comment]
Thanks for answer.

Was just curious why it merged the "./**/*.ts" when my tsconfig.json had no ```include``` or ```files```
Exclude did the job.

Just a little note, we are compiling with TypeScript 2.1.4 already and its working perfectly! But we explicitelly use a dependency in package.json.

**Comment by nycdotnet** [comment]
Hi All,

I have just published `grunt-ts@6.0.0-beta.10` to npm.  It has several improvements for compatibility with both grunt-style stuff (`src`, `files`, etc.) and `tsconfig.json`, with the behavior of `files`, `include`, and `exclude` implemented as TypeScript 2.1 implements it.  Would you please give it a try?

I will close this issue as the original opener seems to have resolved their problem, and it sounds like @Rouche should be in OK shape now too.

Thank you.

**Comment by ikuriel** [comment]
this error also happened to me because i was ussing TypeScript 2.1.0, upgrading to 2.4.x solved it.
